### ISSUE TITLE
Issues downloading large files with low bandwidth with curl against OSD deployed router

### ISSUE THREAD

**Issue by rccrdpccl** [openshift/router#448]
We have encountered an issue when downloading large files with limited bandwidth with curl against OSD based api.openshift.com with HTTP2.

The problem manifests itself as follows:

```bash
podman run curlimages/curl:7.65.3 curl --limit-rate 10M -vvv 'https://api.openshift.com/api/assisted-images/boot-artifacts/rootfs?arch=x86_64&version=4.12' -o /tmp/bla
% Total % Received % Xferd Average Speed Time Time Time Current
Dload Upload Total Spent Left Speed
97 1052M 97 1024M 0 0 9.9M 0 0:01:45 0:01:42 0:00:03 10.1M
curl: (18) transfer closed with 29821952 bytes remaining to read 
```

The download always stops at 1GB (1024^3 bytes exactly)

This is true with all versions of curl up to 7.68 included, however we have noticed that from curl 7.69 this _seems_ to not happen.
With the help of RHEL curl maintainer we found out that in this version the initial window size would change from 1<<30 (which is the same as 1024^3) to 32MB (32 * 1024 * 1024)
https://bugzilla.redhat.com/show_bug.cgi?id=2166254

Surely enough, if we rate limit enough this behaviour occurs with curl 7.69 or greater

```bash
curl -A "banana" --limit-rate 538000 --output /dev/null 'https://api.openshift.com/api/as
sisted-images/boot-artifacts/rootfs?arch=x86_64&version=4.12'
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  3 1052M    3 32.0M    0     0   540k      0  0:33:13  0:01:00  0:32:13  784k
curl: (18) transfer closed with 1070009344 bytes remaining to read
```

And the difference it's again `total_image_size - 32MB`.
Interestigly, trying to find the maximum speed at which the transfer would fail, we noticed that if we're on the limit sometimes the download would fail at what it looks like to be the second window:

```bash
~ $ curl -A "banana" --limit-rate 538000 --output /dev/null 'https://api.openshift.com/api/as
sisted-images/boot-artifacts/rootfs?arch=x86_64&version=4.12'
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  4 1052M    4 48.0M    0     0   530k      0  0:33:50  0:01:32  0:32:18  501k
curl: (18) transfer closed with 1053230365 bytes remaining to read
```

In this case above it seems that they have negotiated a second window of half the size of the first one, but still errored at the end of the second window.

The above behaviour seems to happen only with the following factors:
* HTTP2 (if we force HTTP1.1, this behaviour won't happen)
* slow download of the window size (for 32MB above 1 minute, and for 1GB seemed to be triggered above 74s)
* files need to be bigger than window size
* only curl tested, but might happen on other client if they have similar window sizes

This does not happen on other HTTP2 servers, like cloudfront (mirror.openshift.com for example)

We have tested local plain haproxy with HTTP2, trying to reproduce the bug, but we were unable to do so

**Comment by rccrdpccl** [comment]
We realize this might also be AWS undesired behaviour, but it would be great if we could discard/confirm this behaviour being produced by openshift router

**Comment by frobware** [comment]
> We have tested local plain haproxy with HTTP2, trying to reproduce the bug, but we were unable to do so

Please could you share the haproxy configuration you used for this standalone test.

**Comment by rccrdpccl** [comment]
Sure - I've added a working example [here](https://github.com/rccrdpccl/examples/tree/master/http2/haproxy). You can find the haproxy config [here](https://github.com/rccrdpccl/examples/blob/master/http2/haproxy/haproxy.cfg)

**Comment by frobware** [comment]
> Sure - I've added a working example [here](https://github.com/rccrdpccl/examples/tree/master/http2/haproxy). You can find the haproxy config [here](https://github.com/rccrdpccl/examples/blob/master/http2/haproxy/haproxy.cfg)

Thanks. Which version of haproxy did you use?

**Comment by rccrdpccl** [comment]
I've used `2.2.24`, we can quickly test other versions by changing [here](https://github.com/rccrdpccl/examples/blob/master/http2/haproxy/docker-compose.yaml#L4)

**Comment by frobware** [comment]
> I've used `2.2.24`, we can quickly test other versions by changing [here](https://github.com/rccrdpccl/examples/blob/master/http2/haproxy/docker-compose.yaml#L4)

I missed that, thanks. 

If you could test latest in the 2.0 series (2.2.28) that would be extremely helpful. 

And, if it still fails, then testing 2.6.8 would also be helpful.

**Comment by rccrdpccl** [comment]
Just to be clear, this test *never* reproduced the bug, I've stated it to indicate that this bug does not seem to come from haproxy.

We've been testing the following versions:
* 2.2.24
* 2.2.25
* 2.2.26
* 2.2.27
* 2.2.28
* 2.6.8

I can confirm the issue is *not happening* in none of those verisons above, with the setup linked in the thread above

**Comment by frobware** [comment]
> it would be great if we could discard/confirm this behaviour being produced by openshift router

> Just to be clear, this test never reproduced the bug, I've stated it to indicate that this bug does not seem to come from haproxy.

> I can confirm the issue is not happening in none of those verisons above, with the setup linked in the thread above

So if I understand - it doesn't come from the haproxy versions you've tried standalone but it does still occur with openshift router?

**Comment by rccrdpccl** [comment]
> So if I understand - it doesn't come from the haproxy versions you've tried standalone but it does still occur with openshift router?

Yes, that is correct.

We're not 100% sure it's the openshift router alone though, as we have in front an AWS ELB that might be playing a role in this.

Unfortunately I do not have access to the exact configuration of this specific openshift router nor I was able to skip the ELB to try to isolate the problem between those two components.

We would like to identify if the issue comes from:
* openshift router
* openshift router + ELB
* ELB

The purpose of this issue would be to confirm/discard the first option. Do you happen to know of any quick way we could achieve this?

**Comment by frobware** [comment]
> We would like to identify if the issue comes from:
> 
> * openshift router

Could you try your curl test by passing a `-H "Host: <hostname>"` header, target the IP address of a router pod and do this from another pod in the cluster. I'm just trying to take the ELB out of the equation.

**Comment by rccrdpccl** [comment]
We did already try this, but for some reason we get connection reset. I am unsure what is causing this:

```bash
~ $ curl --resolve "api.openshift.com:443:<redacted>" --resolve "api.openshift.com:80:<redacted>" 'htt
ps://api.openshift.com/api/assisted-images/boot-artifacts/rootfs?arch=x86_64&version=4.12'
curl: (35) OpenSSL SSL_connect: Connection reset by peer in connection to api.openshift.com:443 
```

```bash
~ $ curl --resolve "api.openshift.com:443:<redacted>" --resolve "api.openshift.com:80:<redacted>" 'htt
p://api.openshift.com/api/assisted-images/boot-artifacts/rootfs?arch=x86_64&version=4.12'
curl: (56) Recv failure: Connection reset by peer
```

Of course same result with `-H "Host: api.openshift.com` and targeting the same (redacted) IP.

Could this behaviour be due to the routers' config?

**Comment by frobware** [comment]
Are there any logs in the backend server/pod? How long does the curl run for before the connection is dropped? 

You could also enable access logging on the router to see if there's a hint/clue as to which component drops the connection. Failing that, you may have to do gather a pcap.

**Comment by frobware** [comment]
> Could this behaviour be due to the routers' config?

Is HTTP/2 enabled for this ingress controller (i.e., router)?

**Comment by rccrdpccl** [comment]
Yes, HTTP2 is enabled and this behaviour (the issue's `transfer closed with 29821952 bytes remaining to read`) happens _only_ with HTTP2: if I force HTTP1.1 this does not happen (this beahviour seems to be related to http2 flow control initial-window-size).

This connection reset however happens with both protocols

There are no relevant logs in the target backend, and curl drops the connection instantly:

```bash
$ curl --write-out 'pretransfer:   %{time_pretransfer}\nstarttransfer: %{time_starttransfer}\ntotal:    
     %{time_total}\n' -vvvv --resolve "api.openshift.com:443:<redacted>" --resolve "api.openshift.com:80
:<redacted>" 'https://api.openshift.com/api/assisted-images/boot-artifacts/rootfs?arch=x86_64&version=4.
12'
* Added api.openshift.com:443:<redacted> to DNS cache
* Added api.openshift.com:80:<redacted> to DNS cache
* Hostname api.openshift.com was found in DNS cache
*   Trying <redacted>:443...
* Connected to api.openshift.com (<redacted>) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*   CAfile: /cacert.pem
  CApath: none
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
* OpenSSL SSL_connect: Connection reset by peer in connection to api.openshift.com:443 
* Closing connection 0
curl: (35) OpenSSL SSL_connect: Connection reset by peer in connection to api.openshift.com:443 
pretransfer:   0.000000
starttransfer: 0.000000
total:         0.012893
```

Unfortunately I do not have access to the openshift-ingress resources.

In general I don't think this connection reset is a problem, because the router works generally, it just stops me from bypassing AWS infra sadly.

**Comment by frobware** [comment]
> Yes, HTTP2 is enabled and this behaviour (the issue's `transfer closed with 29821952 bytes remaining to read`) happens _only_ with HTTP2: if I force HTTP1.1 this does not happen (this beahviour seems to be related to http2 flow control initial-window-size).

I just wanted to clarify that HTTP/2 is enabled on the ingress controller as opposed to forcing HTTP/2 using `curl --http2`, for example.

https://docs.openshift.com/container-platform/4.9/networking/ingress-operator.html#nw-http2-haproxy_configuring-ingress

**Comment by rccrdpccl** [comment]
Yes, I can confirm it is enabled on the ingress controller, no need to force it.
```bash
 $ curl -I https://api.openshift.com
HTTP/2 200 
```
